feature/firmware/src/main/kotlin/org/meshtastic/feature/firmware/FirmwareUpdateViewModel.kt

```
@@ -0,0 +1,493 @@
+            viewModelScope.launch {
+                _state.value = FirmwareUpdateState.Checking
+
+                runCatching {
+                    val validationResult = validateDeviceAndConnection()
+
+                    if (validationResult == null) {
+                        // Validation failed, state is already set to Error inside validateDeviceAndConnection
+                        return@launch
+                    }
+
+                    val (ourNode, _, address) = validationResult
+                    val deviceHardware = getDeviceHardware(ourNode) ?: return@launch
+
+                    firmwareReleaseRepository.getReleaseFlow(_selectedReleaseType.value).collectLatest { release ->
+                        _state.value = FirmwareUpdateState.Ready(release, deviceHardware, address)
+                    }
+                }
+                    .onFailure { e ->
+                        if (e is CancellationException) throw e
+                        Timber.e(e)
+                        _state.value = FirmwareUpdateState.Error(e.message ?: "Unknown error")
+                    }
+            }
+    }
+
+    /**
+     * Starts the firmware update process using the currently identified release.
+     * 1. Downloads the firmware zip from the release URL.
+     * 2. Extracts the correct firmware image for the connected device hardware.
+     * 3. Initiates the DFU process.
+     */
+    @Suppress("TooGenericExceptionCaught")
+    fun startUpdate() {
+        val currentState = _state.value as? FirmwareUpdateState.Ready ?: return
+        val (release, hardware, address) = currentState
+
+        if (release == null || !isValidBluetoothAddress(address)) return
+
+        updateJob?.cancel()
+        updateJob =
+            viewModelScope.launch {
+                try {
+                    // 1. Download
+                    _state.value = FirmwareUpdateState.Downloading(0f)
+                    val zipUrl = release.zipUrl.replace("esp32", "nrf52840")
```

Couple questions here : 

- What if the URL doesn't contain "esp32"? It may silently continues with the wrong firmware
  - What if "esp32" appears multiple times in the URL? Only the first occurrence is replaced
  - This assumes the release structure for nrf52840 matches esp32, which may not be true

Users could flash completely wrong firmware to their devices

May need to add some defensive checks here